src/HDF5.jl

```diff
@@ -1752,23 +1752,35 @@ write(parent::Union{HDF5File, HDF5Group}, name::String, data::Union{T, AbstractA
 # For datasets, "write(dset, name, val)" means "a_write"
 write(parent::HDF5Dataset, name::String, data::Union{T, AbstractArray{T}}, plists...) where {T<:ScalarOrString} = a_write(parent, name, data, plists...)
 
-# Reading arrays using getindex: data = dset[:,:,10]
-function getindex(dset::HDF5Dataset, indices::Union{AbstractRange{Int},Int}...)
+
+# Indexing
+
+Array(dset::HDF5Dataset) = read(dset)
+
+Base.eachindex(::IndexLinear, A::HDF5Dataset) = Base.OneTo(length(A))
+Base.axes(dset::HDF5Dataset) = map(Base.OneTo, size(dset))
+
+getindex(dset::HDF5Dataset, I::Union{AbstractRange{Int},Int,Colon}...) = begin
```

Comment:
```
function getindex(dset::HDF5Dataset, I...)
    _getindex(dset, Base.to_indices(dset, I)...)
end
```

we want to be able to index with UInt or Int32 , the method restriction is necessary.

Reply:
I'm also not confident about the bounds checking approach you took, in fact it doesn't actually do any bounds checking. 
We'd need something like
```
@boundscheck checkbounds(A, I...)
```
if you look `hyperslab` is doing to bounds checking right now.  Maybe just leave this to another PR?